src/riscv/ffi.c

```diff
@@ -0,0 +1,746 @@
+                memcpy(argp, *p_argv, z);
+            else
+            {
+                unsigned long portion = cap - (unsigned long)argp;
+                
+                memcpy(argp, *p_argv, portion);
+                argp = stack;
+                z -= portion;
+                memcpy(argp, (void*)((unsigned long)(*p_argv) + portion), z);
+            }
+        }
+        else if(i < 8 && z > 2*sizeof(ffi_arg))
+        {
+            /* It's too big to pass in any registers or on the stack, 
+               so we pass a pointer, and copy the struct to pass by value.
+               But, we can't _just_ copy it onto the stack! We need to actually
+               make sure it gets onto the "bottom" (really the top, high memory
+               addresses) of the stack frame... */
+            
+            /* Update pointer to where our struct location on the stack is */
+            cpy_struct -= ALIGN(z, a);
+            
+            memcpy(cpy_struct, *p_argv, z);
+            
+            /* Pass pointer in register */
+            *(ffi_arg *)argp = (ffi_arg) cpy_struct;
+            
+            z = sizeof(ffi_arg);
+        }
+        else
+        {
+            /* Just some big struct, pass it by value by copying it onto
+               the stack. */
+            memcpy(argp, *p_argv, z);
+        }
+        
+        p_argv++;
+        argp += z;
+    }
+}
+
+/* This code traverses structure definitions 
+   and generates the appropriate flags. */
+
+static unsigned calc_riscv_struct_flags(int soft_float, ffi_type *arg, size_t size, unsigned *loc, unsigned *arg_reg)
+{
+    unsigned flags = 0;
+    unsigned index = 0;
+    ffi_type *e;
+    
+    if (soft_float)
+        return 0;
+    
+    /* The struct is too big to pass on the stack, so we pass it by reference */
+    if (size > 2 * FFI_SIZEOF_ARG)
+    {
+        (*arg_reg)++;
+        return 0;
+    }
+    
+    while ((e = arg->elements[index]))
+    {
+        /* Align this object. */
+        *loc = ALIGN(*loc, e->alignment);
+        
+        if (e->type == FFI_TYPE_DOUBLE)
+        {
+            /* Already aligned to FFI_SIZEOF_ARG. */
+            *arg_reg = *loc / FFI_SIZEOF_ARG;
+            
+            if (*arg_reg > 7)
+                break;
+            
+            flags += (FFI_TYPE_DOUBLE << (*arg_reg * FFI_FLAG_BITS));
+            *loc += e->size;
+        }
+        else
+            *loc += e->size;
+        
+        index++;
+    }
+    
+    /* Next Argument register at alignment of FFI_SIZEOF_ARG. */
+    *arg_reg = ALIGN(*loc, FFI_SIZEOF_ARG) / FFI_SIZEOF_ARG;
+    
+    return flags;
+}
+
+/* The flags output of this routine should match the various struct cases
+   described in ffitarget.h */
+
+static unsigned calc_riscv_return_struct_flags(int soft_float, ffi_type *arg)
+{
+    unsigned flags = 0;
+    unsigned small = FFI_TYPE_SMALLSTRUCT;
+    ffi_type *e;
+    
+    /* Returning structures under n32 is a tricky thing.
```

Comment:
I suspect this comment is a leftover from the MIPS port and that the RISC-V ABI isn't also called "n32".  If so this could use an update.